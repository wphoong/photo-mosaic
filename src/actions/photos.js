import uuid from "uuid";
import database from "../firebase/firebase.js";

// ADD PHOTO

export const addPhoto = (photo) => ({
  type: "ADD_PHOTO",
  photo
});

export const startAddPhoto = (photoData = {}) => {
  return (dispatch) => {
    const {
      photoLink = "",
      title = "",
      createdAt = 0,
      description = ""
    } = photoData;

    const photo = { photoLink, title, createdAt, description };

    return database.ref(`photos`).push(photo).then((ref) => {
      dispatch(addPhoto({
        id: ref.key,
        ...photo
      }));
    });
  };
};

export const editPhoto = (id, updates) => ({
  type: "EDIT_PHOTO",
  id,
  updates
});

export const startEditPhoto = (id, updates) => {
  return (dispatch) => {
    return database.ref(`photos/${id}`).update(updates).then(() => {
      dispatch(editPhoto(id, updates));
    });
  };
};

export const removePhoto = ({id} = {}) => ({
  type: "REMOVE_PHOTO",
  id
});

export const startRemovePhoto = ({id}) => {
  return (dispatch) => {
    return database.ref(`photos/${id}`).remove().then(() => {
      dispatch(removePhoto({id}));
    });
  };
};


export const setPhotos = (photos) => ({
  type: "SET_PHOTOS",
  photos
});

export const startSetPhotos = () => {
  return (dispatch) => {
    return database.ref("photos").once("value").then((snapshot) => {
      const photos = [];

      snapshot.forEach((childSnapshot) => {
        photos.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setPhotos(photos));
    });
  };
};

