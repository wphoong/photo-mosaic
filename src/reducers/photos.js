const photosReducerDefaultState = [];

const photosReducer = (state = photosReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_PHOTO":
      return [
        ...state,
        action.photo
      ];
    default:
      return state;
  }
};

export default photosReducer;