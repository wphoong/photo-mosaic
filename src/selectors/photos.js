import moment from "moment";

const getVisiblePhotos = (photos, { text, sortBy, startDate, endDate}) => {
  return photos.filter((photo) => {
    const createdAtMoment = moment(photos.createdAt);
    const setTextMatch = photo.title.toLowerCase().includes(text.toLowerCase());
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;

    return setTextMatch && startDateMatch && endDateMatch;
  }).sort((a, b) => {
      if (sortBy === "descend") {
        return b.createdAt - a.createdAt;
      } else {
        return a.createdAt - b.createdAt;
      }
    }
  );
};

export default getVisiblePhotos;
