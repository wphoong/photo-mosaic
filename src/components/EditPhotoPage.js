import React from "react";
import { connect } from "react-redux";
import PhotoForm from "./PhotoForm.js";
import { startEditPhoto, startRemovePhoto } from "../actions/photos.js";

export class EditPhotoPage extends React.Component {
  onSubmit = (photo) => {
    this.props.startEditPhoto(this.props.photo.id, photo);
    this.props.history.push("/");
  };
  onRemove = () => {
    this.props.startRemovePhoto({id: this.props.photo.id});
    this.props.history.push("/");
  };
  render () {
    return (
      <div>
        <PhotoForm 
          photo={this.props.photo} 
          onSubmit={this.onSubmit}
        />
        <button
          onClick={this.onRemove}
        >
          Remove
        </button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  photo: state.photos.find((photo) => {
    return photo.id === props.match.params.id;
  })
});

const mapStateToDispatch = (dispatch, props) => ({
  startEditPhoto: (id, photo) => dispatch(startEditPhoto(id, photo)),
  startRemovePhoto: (data) => dispatch(startRemovePhoto(data))
});

export default connect(mapStateToProps, mapStateToDispatch)(EditPhotoPage);

