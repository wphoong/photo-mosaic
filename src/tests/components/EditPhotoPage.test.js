import React from "react";
import { shallow } from "enzyme";
import { EditPhotoPage } from "../../components/EditPhotoPage.js";
import photos from "../fixtures/photos.js";

let wrapper, startEditPhoto, startRemovePhoto, history;

beforeEach(() => {
  startEditPhoto = jest.fn();
  startRemovePhoto = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditPhotoPage 
      photo={photos[0]}
      startEditPhoto={startEditPhoto} 
      startRemovePhoto={startRemovePhoto} 
      history={history}
    />);
});

test("should render edit photo page", () =>{
  expect(wrapper).toMatchSnapshot();
});

test("should handle on submit for edit expense", () => {
  wrapper.find("PhotoForm").prop("onSubmit")(photos[0]);
  expect(startEditPhoto).toHaveBeenLastCalledWith(photos[0].id, photos[0]);
  expect(history.push).toHaveBeenLastCalledWith("/");
});

test("should handle on click for remove expense", () => {
  wrapper.find("button").simulate("click");
  expect(startRemovePhoto).toHaveBeenLastCalledWith({ id: photos[0].id});
  expect(history.push).toHaveBeenLastCalledWith("/");
});
