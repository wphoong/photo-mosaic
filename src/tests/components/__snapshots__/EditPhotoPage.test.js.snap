// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit photo page 1`] = `
<div>
  <PhotoForm
    onSubmit={[Function]}
    photo={
      Object {
        "createdAt": 0,
        "description": "descript1",
        "id": "1",
        "photoLink": "ez",
        "title": "test1",
      }
    }
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
